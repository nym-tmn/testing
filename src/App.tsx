import { useState } from "react"
import { Button } from "./components/Button/Button"
import { LoginModal } from "./components/LoginModal/LoginModal"
import { useAuth } from "./hooks/isAuth/useAuth";

const App = () => {

	const { isAuth, logout } = useAuth();

	const [isOpenModal, setIsOpenModal] = useState(false)

	return (
		<div data-testid='app'>
			{isAuth
				? <>
					<p>Вы авторизованы!!!</p>
					<Button onClick={logout}>Выйти</Button>
				</>
				: <Button onClick={() => setIsOpenModal(true)}>
					Авторизоваться
				</Button>
			}
			<LoginModal isOpenModal={isOpenModal} handleClose={() => setIsOpenModal(false)} />
		</div>
	)
}

export default App
