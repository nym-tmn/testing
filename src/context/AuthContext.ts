import { createContext } from "react";

interface AuthContextType {
	isAuth: boolean;
	login: () => void;
	logout: () => void;
}

export const AuthContext = createContext<AuthContextType | undefined>(undefined);