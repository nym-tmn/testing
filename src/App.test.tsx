import { describe, expect, it, vi } from "vitest";
import { render, screen } from '@testing-library/react';
import App from './App';
import { AuthContext } from "./context/AuthContext";
import userEvent from "@testing-library/user-event";

describe('App component', () => {
	it('Render of a button with the text "Login" if not logged in', () => {

		const login = vi.fn();
		const logout = vi.fn();

		render(
			<AuthContext.Provider value={{ isAuth: false, login, logout }}>
				<App />
			</AuthContext.Provider>);

		expect(screen.getByRole('button', { name: /^Авторизоваться$/ })).toBeInTheDocument();
		expect(screen.queryByText('Вы авторизованы!!!')).toBeNull();
		expect(screen.queryByText('Выйти')).toBeNull();
		expect(screen.getByTestId('app')).toMatchSnapshot();
	});

	it('Render greeting and button with text "Exit" if authorized', () => {

		const login = vi.fn();
		const logout = vi.fn();

		render(
			<AuthContext.Provider value={{ isAuth: true, login, logout }}>
				<App />
			</AuthContext.Provider>);

		expect(screen.getByText(/^Вы авторизованы!!!$/)).toBeInTheDocument()
		expect(screen.getByRole('button', { name: /^Выйти$/ })).toBeInTheDocument();
		expect(screen.queryByText('Авторизоваться')).toBeNull();
		expect(screen.getByTestId('app')).toMatchSnapshot();
	});

	it('should open modal when clicked button "Authorization"', async () => {

		const login = vi.fn();
		const logout = vi.fn();

		render(
			<AuthContext.Provider value={{ isAuth: false, login, logout }}>
				<App />
			</AuthContext.Provider>
		);

		expect(screen.queryByTestId('login-modal')).toBeNull();

		await userEvent.click(screen.getByRole('button', { name: /^Авторизоваться$/ }));

		expect(screen.getByTestId('login-modal')).toBeInTheDocument();
	});

	it('should close modal when clicked button "X"', async () => {

		const login = vi.fn();
		const logout = vi.fn();

		render(
			<AuthContext.Provider value={{ isAuth: false, login, logout }}>
				<App />
			</AuthContext.Provider>
		);

		await userEvent.click(screen.getByRole('button', { name: /Авторизоваться/i }));

		await userEvent.click(screen.getByRole('button', { name: /^X$/ }));

		expect(screen.queryByTestId('login-modal')).toBeNull();
	});
})