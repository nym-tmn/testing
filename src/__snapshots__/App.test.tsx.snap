// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > Render greeting and button with text "Exit" if authorized 1`] = `
<div
  data-testid="app"
>
  <p>
    Вы авторизованы!!!
  </p>
  <button
    class="_button_0829c1"
    data-testid="click-btn"
    type="button"
  >
    Выйти
  </button>
</div>
`;

exports[`App component > Render of a button with the text "Login" if not logged in 1`] = `
<div
  data-testid="app"
>
  <button
    class="_button_0829c1"
    data-testid="click-btn"
    type="button"
  >
    Авторизоваться
  </button>
</div>
`;
